File: Practice/Jdeveloper/javaproject/trie/Trie.java
package javaproject.trie;

public class Trie{
  TrieNode root;

  public Trie(){
    root = new TrieNode();
  }
/* TrieNode
*/
class TrieNode{
  TrieNode[] children;
  boolean isLeaf;

  public TrieNode(){
    this.isLeaf = false;
    children = new TrieNode[26];
  }
}

    public static void main(String[] args){
        Trie t = new Trie();
        t.insert("vinod");
        t.insert("vis");
    }
    
  public void insert(String strValue){
      if(strValue == null || strValue.length() == 0)
        return;

      int length = strValue.length();
      TrieNode node = root;
      for(int i=0;i<length;i++){
        int index = (strValue.charAt(i) - 'a');
        if(node.children[index] == null)
           node.children[index] = new TrieNode();
        node = node.children[index];   
      }

      node.isLeaf = true;
  }

  public boolean search(String strValue){
      if(strValue == null || strValue.length() == 0)
        return false;
      int length = strValue.length();
      TrieNode node = root;
      for(int i=0;i<length;i++){
          int index = (strValue.charAt(i) - 'a');
          if(node.children[index] == null)
              return false;
          node = node.children[index];
      }
    
    return node.isLeaf;
  }
   
  public boolean startWith(String strValue){
      if(strValue == null || strValue.length() == 0)
        return false;
      int length = strValue.length();
      TrieNode node = root;
      for(int i=0;i<length;i++){
          int index = (strValue.charAt(i) - 'a');
          if(node.children[index] == null)
              return false;
          node = node.children[index];
      }
      return true;
  }
    
}

File: Problems/Trie/trie.js
(function(exports){
    'use strict';

  console.clear();

  var Trie = function(){
    this.child = new TrieNode("");
  }

  var TrieNode = function(value){
    this.kids = new Map();
  };

  var list = [
    "abacus",
    "bar",
    "bart",
    "build",
    "cat",
    "cart"
  ];

  var insert = function(list){
    var trieRoot = new Trie();

    //iterate the list
    list.forEach(function(item){

      //insert into it
      insertIntoTrie(item,trieRoot);
    });

    return trieRoot;
  }

  var insertIntoTrie = function(item,root){
      var current = root;

      //iterate the character of the single word
      for(var i=0;i<item.length;i++){

        //get the character
        var ch = item[i];
        //create a trie node
        var child = new TrieNode(ch);

        //See if the kids contains character
        var value = child.kids.get(ch);

        if(value !== null){
          //get the child to kids value
          child = value;
        }
        else{
          //set the kids value
          child.kids.set(ch, new TrieNode(ch));
        }
        //assign current value
        current = child;
      }
  }

  var search  =function(key,root){
    var current = root;
    var child = null;
    for(var i=0;key.length;i++){
      var ch = key[i];
      var value = current.child.kids.get(ch);
       if(value !== null){
         child = value;
       }
       else{
         return false;
       }
      current = child;
    }
  }

  var root = insert(list);
  var exist = search("bart",root);
  console.log(exist);

})(this)
